hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/clean/CleanPlanActionExecutor.java

```diff
@@ -86,7 +92,29 @@ private boolean needsCleaning(CleaningTriggerStrategy strategy) {
     if (strategy == CleaningTriggerStrategy.NUM_COMMITS) {
       int numberOfCommits = getCommitsSinceLastCleaning();
       int maxInlineCommitsForNextClean = config.getCleaningMaxCommits();
-      return numberOfCommits >= maxInlineCommitsForNextClean;
+      if (numberOfCommits >= maxInlineCommitsForNextClean) {
+        // check if the number of commits created after the last clean is greater than clean.max.commits
+        int commitsRetained = config.getCleanerCommitsRetained();
+        int hoursRetained = config.getCleanerHoursRetained();
+        if (config.getCleanerPolicy() == HoodieCleaningPolicy.KEEP_LATEST_COMMITS) {
+          // if cleaner policy is KEEP_LATEST_COMMITS then
+          // check if the number of completed commits in the timeline is greater than cleaner.commits.retained
+          return table.getCompletedCommitsTimeline().countInstants() > commitsRetained;
```

Comment:
These checks are already present in CleanPlanner here - https://github.com/apache/hudi/blob/a0df6ecfb47ac9e393b5eb977850cb8fbe7d3f72/hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/clean/CleanPlanner.java#L313. 

The new changes are redundant and we should remove the checks from any one place. I will leave this to you to decide. I feel we should remove the check from the place I mentioned since this makes more sense to keep them in this method here.